@xen-orchestra/proxy/src/app/_debounceResource.js

```diff
@@ -1,35 +1,53 @@
+import asyncMapSettled from '@xen-orchestra/async-map'
 import Disposable from 'promise-toolbox/Disposable'
+import { createLogger } from '@xen-orchestra/log/dist'
 
-export async function debounceResource(pDisposable, hooks, delay = 0) {
-  if (delay === 0) {
-    return pDisposable
-  }
+const { warn } = createLogger('xo:proxy:debounceResource')
+
+export const createDebounceResource = defaultDelay => {
+  const flushers = new Set()
+  async function debounceResource(pDisposable, delay = defaultDelay) {
+    if (delay === 0) {
+      return pDisposable
+    }
 
-  const disposable = await pDisposable
+    const disposable = await pDisposable
 
-  let timeoutId
-  const disposeWrapper = () => {
-    if (timeoutId !== undefined) {
-      clearTimeout(timeoutId)
-      timeoutId = undefined
-      hooks.removeListener('stop', onStop)
+    let timeoutId
+    const disposeWrapper = async () => {
+      if (timeoutId !== undefined) {
+        clearTimeout(timeoutId)
+        timeoutId = undefined
+        flushers.delete(disposer)
 
-      return disposable.dispose()
+        try {
+          await disposable.dispose()
+        } catch (error) {
+          warn(error)
+        }
+      }
     }
-  }
 
-  const onStop = () => {
-    const shouldDisposeNow = timeoutId !== undefined
-    if (shouldDisposeNow) {
-      return disposeWrapper()
-    } else {
-      // will dispose ASAP
-      delay = 0
+    const disposer = () => {
+      const shouldDisposeNow = timeoutId !== undefined
+      if (shouldDisposeNow) {
+        return disposeWrapper()
+      } else {
+        // will dispose ASAP
+        delay = 0
+      }
     }
+    flushers.add(disposer)
+
+    return new Disposable(disposable.value, () => {
+      timeoutId = setTimeout(disposeWrapper, delay)
+    })
+  }
+  debounceResource.flushAll = () => {
+    const currentFlushers = [...flushers]
+    flushers.clear()
+    return asyncMapSettled(currentFlushers, dispose => dispose())
```

Comment:
If flush cannot reject maybe we could use `asyncMap` which works with iterables instead?